trajectory_executor/com_serial.py

```diff
@@ -0,0 +1,222 @@
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_AT_GOAL']
+        elif read_state == serial_read_states['READ_AT_GOAL']:
+          data_packet['at_goal'] = decoded_data
+          calculated_checksum += decoded_data
+          read_state = serial_read_states['READ_CHECKSUM']
+        elif read_state == serial_read_states['READ_CHECKSUM']:
+          data_packet['checksum'] = decoded_data
+          if data_packet['checksum'] == 255 - (calculated_checksum % 256):
+            # good packet, relay to tracking node
+            response.roll = data_packet['roll']
+            response.pitch = data_packet['pitch']
+            response.yaw = data_packet['yaw']
+            response.l_hip = data_packet['l_hip']
+            response.l_knee = data_packet['l_hip']
+            response.r_hip = data_packet['r_hip']
+            response.r_knee = data_packet['r_hip']
+            response.l_shoulder = data_packet['l_shoulder']
+            response.l_elbow = data_packet['l_elbow']
+            response.r_shoulder = data_packet['r_shoulder']
+            response.r_elbow = data_packet['r_elbow']
+            response.at_goal = data_packet['at_goal']
+            break
+          else:
+            # bad packet, retry request
+            return self.sensor_data_request_callback(request, response)
+
+    return response
+
+  def send_joint_angles_callback(self, msg: JointAngles):
+    """Send new joint angles to Arduino for execution"""
+    # Split joint angles into two numbers since max value of a byte is 255
+    ja_bytes = []
+    ja_bytes.append(255 if msg.left_hip // 256 > 0 else msg.left_hip)
+    ja_bytes.append(msg.left_hip % 255 if msg.left_hip // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.left_knee // 256 > 0 else msg.left_knee)
+    ja_bytes.append(msg.left_knee % 255 if msg.left_knee // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_hip // 256 > 0 else msg.right_hip)
+    ja_bytes.append(msg.right_hip % 255 if msg.right_hip // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_knee // 256 > 0 else msg.right_knee)
+    ja_bytes.append(msg.right_knee % 255 if msg.right_knee // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.left_shoulder // 256 > 0 else msg.left_shoulder)
+    ja_bytes.append(msg.left_shoulder % 255 if msg.left_shoulder // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.left_elbow // 256 > 0 else msg.left_elbow)
+    ja_bytes.append(msg.left_elbow % 255 if msg.left_elbow // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_shoulder // 256 > 0 else msg.right_shoulder)
+    ja_bytes.append(msg.right_shoulder % 255 if msg.right_shoulder // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_elbow // 256 > 0 else msg.right_elbow)
+    ja_bytes.append(msg.right_elbow % 255 if msg.right_elbow // 256 > 0 else 0)
```

Comment:
again, not 100% sure on how ROS creates its data objects, but this could potentially be reduced to:

```python
joints = (attr for attr in dir(msg) if not attr.startswith('_'))
ja_bytes = []
for joint in joints:
  if getattr(msg, joint) // 256 > 0:
    ja_bytes.append(255)
    ja_bytes.append(getattr(msg, joint) % 255)
  else:
    ja_bytes.append(getattr(msg, joint))
    ja_bytes.append(0)
```

Reply:
Yea so this one I did look into a bit because it's really ugly and I hate it but so far it seems that there isn't a way to iterate over the items of a ROS message like you would do with a python dictionary. One way to get around this would be to have the message object only have one field: a list of numbers representing the joint angles. I avoided this solution for the first run just because we lose explicit assignment of values to their respective joints and rely on the list order being maintained during transmission. Purely documentation reasons. I'm ok with switching the message format to a list if you guys are ok with it and really want to get rid of this unholy mess.

Reply:
I hate this holy mess but I think we have other important priorities. I am okay with creating a bunch of code improvement related issues so that we can resolve them. I don't think it's wise to spend time on this right this moment

Reply:
Yep. We'll create a ticket in trello for this